lib/puppet/functions/regsubst.rb

```diff
@@ -95,7 +95,27 @@ def regsubst_regexp(target, pattern, replacement, flags = nil)
   end
 
   def inner_regsubst(target, re, replacement, op)
-    target.respond_to?(op) ? target.send(op, re, replacement) : target.collect { |e| e.send(op, re, replacement) }
+    if target.respond_to?(:map)
+      # this is an Array
+      target.map do |item|
+        if item.respond_to?(:unwrap)
+          item = item.unwrap
+          Puppet::Pops::Types::PSensitiveType::Sensitive.new(
+            item.respond_to?(op) ? item.send(op, re, replacement) : item.collect { |e| e.send(op, re, replacement) }
+          )
+        else
+          item.respond_to?(op) ? item.send(op, re, replacement) : item.collect { |e| e.send(op, re, replacement) }
+        end
```

Comment:
Could the array case be handled recursively instead of copy/pasta?

```ruby
if target.is_a?(Array)
  target.map do |item|
    inner_regsubst(item, re, replacement, op)
  end
elsif target.is_a?(Puppet::Pops::Types::PSensitiveType::Sensitive)
```